demo/src/ddnet/writer.rs

```diff
@@ -0,0 +1,205 @@
+        let type_id = match self.index.entry(*uuid) {
+            hash_map::Entry::Occupied(type_id) => *type_id.get(),
+            hash_map::Entry::Vacant(uuid_entry) => {
+                let type_id = self.next_type_id;
+                uuid_entry.insert(type_id);
+                self.next_type_id -= 1;
+                type_id
+            }
+        };
+        if !self.advertised.contains(uuid) {
+            self.advertised.insert(*uuid);
+            let mut uuid_item_data = [0; 4];
+            for (mut_data, uuid_data) in uuid_item_data.iter_mut().zip(uuid.chunks(4)) {
+                *mut_data = i32::from_be_bytes(uuid_data.try_into().unwrap());
+            }
+            builder.add_item(0, type_id, &uuid_item_data)?;
+        }
+        Ok(type_id)
+    }
+}
+
+fn mem_swap_default<T: Default>(mut_ref: &mut T) -> T {
+    let mut swap = T::default();
+    mem::swap(&mut swap, mut_ref);
+    swap
+}
+
+impl DemoWriter {
+    pub fn new<T: io::Write + io::Seek + 'static>(
+        file: T,
+        net_version: &[u8],
+        map_name: &[u8],
+        map_sha256: Option<common::digest::Sha256>,
+        map_crc: u32,
+        kind: crate::DemoKind,
+        length: i32,
+        timestamp: &[u8],
+        map: &[u8],
+    ) -> Result<Self, WriteError> {
+        let raw = crate::Writer::new(
+            file,
+            net_version,
+            map_name,
+            map_sha256,
+            map_crc,
+            kind,
+            length,
+            timestamp,
+            map,
+        )?;
+
+        Ok(Self {
+            inner: raw,
+            last_tick: -1,
+            last_keyframe: None,
+            uuid_index: UuidIndex::new(),
+            snap: snapshot::Snap::default(),
+            delta: snapshot::Delta::default(),
+            builder: snap::Builder::default(),
+            buf: arrayvec::ArrayVec::new(),
+            i32_buf: Vec::new(),
+        })
+    }
+
+    pub fn write_snap<'a, T: Iterator<Item = &'a (snap_obj::SnapObj, u16)>>(
+        &mut self,
+        tick: i32,
+        items: T,
+    ) -> Result<(), WriteError> {
+        if tick < self.last_tick {
+            return Err(WriteError::TooLowTickNumber);
+        }
+        let is_keyframe = match self.last_keyframe {
+            None => true,
+            Some(last_keyframe) => tick - last_keyframe >= 125,
+        };
+
+        self.uuid_index.reset();
+        for (item, id) in items {
+            let type_id = match item.obj_type_id() {
+                snap_obj::TypeId::Ordinal(type_id) => type_id,
+                snap_obj::TypeId::Uuid(uuid) => {
+                    let uuid = uuid.as_bytes();
+                    self.uuid_index.get_type_id(&mut self.builder, uuid)?
+                }
+            };
+            self.builder.add_item(type_id, *id, item.encode())?;
+        }
+
+        let old_snap = mem_swap_default(&mut self.snap);
+        let new_snap = mem_swap_default(&mut self.builder).finish();
```

Comment:
```suggestion
        let new_snap = mem::replace(&mut self.builder, Default::default()).finish();
```
Same for the line above.

Reply:
Uses `mem::take` now